crates/core/src/vm.rs

```diff
@@ -328,26 +354,32 @@ impl<'db, 'tx> DbProgram<'db, 'tx, MutTx> {
     }
 
     fn _execute_insert(&mut self, table: &Table, rows: Vec<ProductValue>) -> Result<Code, ErrorVm> {
-        match table {
-            // TODO: How do we deal with mutating values?
-            Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
-            Table::DbTable(x) => {
-                for row in rows {
-                    self.db.insert(self.tx, x.table_id, row)?;
+        match self.tx {
+            TxMode::MutTx(tx) => match table {
+                // TODO: How do we deal with mutating values?
+                Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
+                Table::DbTable(x) => {
+                    for row in rows {
+                        self.db.insert(tx, x.table_id, row)?;
+                    }
+                    Ok(Code::Pass)
                 }
-                Ok(Code::Pass)
-            }
+            },
+            TxMode::Tx(_) => Err(ErrorVm::Unsupported("Read tx".to_owned())),
         }
     }
 
     fn _execute_delete(&mut self, table: &Table, rows: Vec<ProductValue>) -> Result<Code, ErrorVm> {
-        match table {
-            // TODO: How do we deal with mutating values?
-            Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
-            Table::DbTable(t) => {
-                let count = self.db.delete_by_rel(self.tx, t.table_id, rows);
-                Ok(Code::Value(count.into()))
-            }
+        match self.tx {
+            TxMode::MutTx(tx) => match table {
+                // TODO: How do we deal with mutating values?
+                Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
+                Table::DbTable(t) => {
+                    let count = self.db.delete_by_rel(tx, t.table_id, rows);
+                    Ok(Code::Value(count.into()))
+                }
+            },
+            TxMode::Tx(_) => Err(ErrorVm::Unsupported("Read tx".to_owned())),
```

Comment:
Is this an error or is this actually unreachable? Like is it possible to have a TxMode::Tx if you've arrived at this function?

Reply:
This should be unreachable code.